### PR TITLE
Adding one-sided MLS support and example for IBMethod

### PR DESCRIPTION
**Pull request by amneetb**

### CODE REVIEW

**Comment by amneetb** [comment]
@boyceg : Can you add first-class support for your favorite kernels here?:
https://github.com/IBAMR/IBAMR/blob/f232e3e59f9d85c4cf0e12d9b0b5d04a9c772a5e/ibtk/src/lagrangian/LEInteractor.cpp#L1133-L1158

File: ibtk/Makefile.in [hunk 1]
```diff
@@ -188,8 +188,7 @@ am__DIST_COMMON = $(srcdir)/Makefile.in \
 	$(top_srcdir)/config/install-sh $(top_srcdir)/config/ltmain.sh \
 	$(top_srcdir)/config/missing config/compile \
 	config/config.guess config/config.rpath config/config.sub \
-	config/depcomp config/install-sh config/ltmain.sh \
-	config/missing
+	config/install-sh config/ltmain.sh config/missing
```
Comment (drwells) [inline ibtk/Makefile.in]: This isn't the most important problem we have but `depcomp` really should be installed. Are you sure that you are using autotools 2.69?

File: ibtk/include/ibtk/LEInteractor.h [hunk 2]
```diff
@@ -433,8 +485,8 @@ class LEInteractor
                             int Q_depth,
                             const std::vector<double>& X_data,
                             int X_depth,
-                            SAMRAI::tbox::Pointer<SAMRAI::pdat::SideData<NDIM, double> > mask_data,
                             SAMRAI::tbox::Pointer<SAMRAI::pdat::SideData<NDIM, double> > q_data,
+                            SAMRAI::tbox::Pointer<SAMRAI::pdat::SideData<NDIM, double> > mask_data,
```
Comment (drwells) [inline ibtk/include/ibtk/LEInteractor.h]: Did this just not work before?

**Comment by boyceg** [comment]
@amneetb, is the example for a fixed geometry? If so, can we hold off on the API changes for `LDataManager` / `LEInteractor` / for now? I hope to be able to do this without doubling the number of function prototypes for spreading and interpolation.

**Comment by amneetb** [comment]
@boyceg: Moving geometries can also be done by the same API. The change in the position gets reflected in the mask_data_idx.

@drwells: I have automake 1.16 and autoconf 2.69 on my system obtained through homebrew. I recently updated all of system including OSX, so everything is at the latest version.

**Comment by boyceg** [comment]
> @boyceg: Moving geometries can also be done by the same API. The change in the position gets reflected in the mask_data_idx.

I am saying: let's figure out the API once we know that this is worth doing.

**Comment by boyceg** [comment]
@drwells this is not really ready to be reviewed -- this is test code to investigate whether an alternative kernel function construction can give better results for certain problems.

File: ibtk/include/ibtk/LEInteractor.h (hunk 2 above)
Comment (amneetb) [reply ibtk/include/ibtk/LEInteractor.h]: I had one-sided MLS for IBFE before  (that was removed from the library). @boyceg wants to revive it and we are doing it for IBMethod first.

**Comment by boyceg** [comment]
> > @boyceg: Moving geometries can also be done by the same API. The change in the position gets reflected in the mask_data_idx.
> 
> I am saying: let's figure out the API once we know that this is worth doing.

For now, I would suggest just having a very simple callback function in `LEInteractor` that lets you specify a function that knows whether you are "inside" or "outside" of the region of interest.

**Comment by amneetb** [comment]
Well, I again spent a lot of time and effort on something that will purged again from the library :-(. I leave this PR to you and Fuhui to decide its worthiness.

**Comment by boyceg** [comment]
> > > @boyceg: Moving geometries can also be done by the same API. The change in the position gets reflected in the mask_data_idx.
> > 
> > 
> > I am saying: let's figure out the API once we know that this is worth doing.
> 
> For now, I would suggest just having a very simple callback function in `LEInteractor` that lets you specify a function that knows whether you are "inside" or "outside" of the region of interest.

(I think that this means that the only interface changes would be to add a way to tell `LEInteractor` about the masking function.)

**Comment by boyceg** [comment]
@amneetb, for the tests that we discussed, I think that there needs to be a simple way to tell the MLS construction where the support of the kernel function should go, to make it easy to try out with different simple geometries.

**Comment by amneetb** [comment]
Yes, but you an Eulerian patch data at the end of the day, no matter whatever you decide to do. I don’t understand why adding a mask_data_idx is not the right thing to do. The user is free to fill it however they feel like. They analytical function (like the example I rolled with this PR), or use FESurfaceDistanceEvalutor class. 

Just take a look at my example!

**Comment by boyceg** [comment]
> Yes, but you an Eulerian patch data at the end of the day, no matter whatever you decide to do.

You don't _need_ to use `PatchData` to specify the support of the kernel -- you just need to know which spatial locations are in the support.

> I don’t understand why adding a mask_data_idx is not the right thing to do. The user is free to fill it however they feel like. They analytical function (like the example I rolled with this PR), or use FESurfaceDistanceEvalutor class.

I agree that you will want to use `PatchData` for complex geometries, but I don't think we are there yet...

The problem with adding the `mask_idx` to the `LEInteractor::spread()` and `LEInteractor::interp()` interfaces is that means that the change has to be propagated throughout the library, which is a lot of work.

Until we know that we actually want to use these kernels, I am suggesting that we minimize the changes by just providing a very simple mechanism to tell the MLS code where the kernels are supported. This will make it extremely easy to try out with CIB as well as with "direct forcing" versions of IB and IBFE without requiring any changes other than creating a function to evaluate the "indicator function" for the specific geometry.

**Comment by amneetb** [comment]
@fuhuifang : To make preconditioner work for a kernel different than `IB_3`, `IB_4`, and `IB_6` change this line of code

https://github.com/IBAMR/IBAMR/blob/amneetb/revive-mls/src/IB/CIBMethod.cpp#L1403

Basically check that if the default kernel is not one of the three (`IB_3`, `IB_4`,  `IB_6`), then in the preconditioner set 

`ib_kernel = "IB_6"`  

as your go-to kernel.

**Pull request closed without merging**